daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Recursion.scala

```diff
@@ -25,6 +25,8 @@ private[validation] object Recursion {
     val modRefsInType: Set[ModuleName] = {
 
       def modRefsInType(acc: Set[ModuleName], typ0: Type): Set[ModuleName] = typ0 match {
+        case TSynApp(typeSynName,_) if typeSynName.packageId == pkgId =>
```

Comment:
Don’t you need to recurse on the arguments here?

Reply:
Wont that happen in the `TypeTraversable`?

Reply:
Ah, no it wont. Think you're right.

Reply:
Actually, take 3. I'm not sure.
Too much scala magic for my brain to process :(

Reply:
My understanding is that `TypeTraversable` is roughly the equivalent of something like Haskell’s `plate`, i.e., a traversal for the immediate children that only goes down one level. Then it’s up to the function that you’re using to traverse to do further recursion, that’s why the `otherwise`case here calls the traversable again. Since you don’t fall into the otherwise case, you need to recurse on the arguments either by inlining the recursion on the arguments or my calling the traversable here as well.

Reply:
`TypeTraversable(typp: Type)` gave you an traversable over the *direct* subtree (of type `Type`) of `typ0`.
If you want to iterate over all the subtree of `typ` you have to recurse on each subtree. 
Here that is done by the default case (line 32/33).
Since `TSynApp(typeSynName,_)` has its own case and contains subtree you have to recurse.

Reply:
:+1: for @cocreature explanation.

Reply:
Right. Thank you. I see now it is not doing deep recursion.